refactor: Remove all policy globals

maflcko opened the pull request:
This change is good because:

* It moves module-specific init-logic out of the bloated init.cpp
* It removes a global from validation.cpp and places it into the data structure that needs it (mempool)

Commit:
Return optional error from ApplyArgsManOptions

Also pass in a (for now unused) reference to the params.

Both changes are needed for the next commit.

Commit:
Remove ::fRequireStandard global

jonatack commented:
Approach ACK

luke-jr commented:
Concept ACK

glozow reviewed (commented):
Approach ACK

`MemPoolOptions` seems like the correct place to put these things since `full_rbf` is already there.

I would put `::incrementalRelayFee`, `::minRelayFee`, `dustRelayFee`, `fIsBareMultisigStd` in the same boat (sorry for scope creep, probably for a future PR).

Commit:
Remove ::incrementalRelayFee and ::minRelayTxFee globals

Commit:
test: Remove unused cs_main

Commit:
Remove ::IsStandardTx(tx, reason) alias

Apart from tests, it is only used in one place, so there is no need for
an alias.

Commit:
Remove ::dustRelayFee

MarcoFalke commented on src/kernel/mempool_options.h:
```diff
@@ -33,6 +36,9 @@ struct MemPoolOptions {
     std::chrono::seconds expiry{std::chrono::hours{DEFAULT_MEMPOOL_EXPIRY_HOURS}};
+    CFeeRate incremental_relay_fee{DEFAULT_INCREMENTAL_RELAY_FEE};
+    /** A fee rate smaller than this is considered zero fee (for relaying, mining and transaction creation) */
+    CFeeRate min_relay_fee{DEFAULT_MIN_RELAY_TX_FEE};
```
Thx, done

Commit:
Remove ::fIsBareMultisigStd global

Commit:
Remove ::GetVirtualTransactionSize() alias

Each alias is only used in one place.

Commit:
Combine datacarrier globals into one

Commit:
Remove ::g_max_datacarrier_bytes global

Commit:
Pass datacarrier setting into IsStandard

MarcoFalke commented:
> sorry for scope creep, probably for a future PR

Thx, all done here

darosior reviewed (commented):
Nice! Concept ACK.

glozow commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -116,14 +116,15 @@ void MockTime(FuzzedDataProvider& fuzzed_data_provider, const CChainState& chain
     // ...override specific options for this specific fuzz suite
     mempool_opts.estimator = nullptr;
     mempool_opts.check_ratio = 1;
+    mempool_opts.require_standard = fuzzed_data_provider.ConsumeBool();
```
in fa3b8a8d7dddbb6fcf51ac447acc0b1c16bbde20
Not changed by this PR, but why don't we just fuzz with require standard always on? false just disables some tests, no?

MarcoFalke replied:
Is there a reason *not* to fuzz the additional branches?

glozow replied:
Not strongly opinionated, resolving this comment, but have a question to clarify my understanding. My thinking here was based on the `check_ratio` set to 1, since any other value just means you run `check()` less often. AFAICT it's the same idea with standardness checks; if you turn it off, you're running fewer tests. Is it another branch if the codepaths you can hit with `true` is a superset of the codepaths you can hit with `false`? Am I misunderstanding?

darosior replied:
I think it is good to also fuzz for non standard transactions, since we support it. Otherwise the validation codepath for non standard transactions that would be executed in "production" would not be fuzzed. On the other hand, "prod" is only testnet here since we don't allow `-acceptnonstdtxn` for mainnet..

nit: Since this is using cycles for a test-only option, maybe we could lower the probability of setting `require_standard` to `false`? Something like `require_standard = (ConsumeIntegralInRange(1, 10) % 10 != 0)`?

MarcoFalke replied:
> AFAICT it's the same idea with standardness checks;

Not really. `check` will not mutate the mempool. However, `require_standard` may leave the mempool in a different state with the same ATMP calls, but a different value of `require_standard`. For example tx-size may create more inputs in the mempool. Or a raw non-standard script may trigger more paths in script evaluation. I think the mempool also uses script caching, so I was hoping for  this to also be a check that the mempool doesn't print coins out of thin air.

glozow replied:
> check will not mutate the mempool. However, require_standard may leave the mempool in a different state with the same ATMP calls, but a different value of require_standar

Ahh yeah, make sense different branches, thanks!

glozow reviewed (commented):
reviewed up to fac940af80a5e8babc9da79843f324147d8e41f5

MarcoFalke commented on src/test/transaction_tests.cpp:
```diff
@@ -745,7 +745,6 @@ BOOST_AUTO_TEST_CASE(test_witness)
 
 BOOST_AUTO_TEST_CASE(test_IsStandard)
 {
-    LOCK(cs_main);
```
Can you explain how it is used?

glozow replied:
oh oops, I thought dustRelayFee needed it for some reason

MarcoFalke commented on src/mempool_args.h:
```diff
@@ -16,7 +20,7 @@ struct MemPoolOptions;
  * @param[in,out] mempool_opts The MemPoolOptions to modify according to \p argsman.
  */
-void ApplyArgsManOptions(const ArgsManager& argsman, kernel::MemPoolOptions& mempool_opts);
+std::optional<bilingual_str> ApplyArgsManOptions(const ArgsManager& argsman, const CChainParams& chainparams, kernel::MemPoolOptions& mempool_opts);
```
Thx, done

MarcoFalke commented on src/validation.cpp:
```diff
@@ -840,7 +841,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
                 strprintf("%d", nSigOpsCost));
 
-    // No individual transactions are allowed below minRelayTxFee and mempool min fee except from
+    // No individual transactions are allowed below the min relay fee and mempool min fee except from
```
Thx, done

MarcoFalke commented:
Rebased

glozow commented on src/mempool_args.cpp:
```diff
@@ -39,6 +44,29 @@ std::optional<bilingual_str> ApplyArgsManOptions(const ArgsManager& argsman, con
+        } else {
+            return AmountErrMsg("minrelaytxfee", argsman.GetArg("-minrelaytxfee", ""));
+        }
+    } else if (mempool_opts.incremental_relay_feerate > mempool_opts.min_relay_feerate) {
+        // Allow only setting incremental fee to control both
+        mempool_opts.min_relay_feerate = mempool_opts.incremental_relay_feerate;
+        LogPrintf("Increasing minrelaytxfee to %s to match incrementalrelayfee\n", mempool_opts.min_relay_feerate.ToString());
+    }
```
in fad51d5dac5b1da1540dd233c6bd096cb3e4a2a5, not changed by this PR, but interesting to note that incremental relay feerate must be set before min relay feerate.

glozow commented on src/script/standard.h:
```diff
@@ -41,11 +42,11 @@ static const unsigned int MAX_OP_RETURN_RELAY = 83;
  * type is designated as TxoutType::NULL_DATA.
+ *
+ * Maximum size of TxoutType::NULL_DATA scripts that this node considers standard.
+ * If nullopt, any size is nonstandard.
```
Agree with the approach of optional<size> in fac85fe0632eace9b49d5bdbc0e9e47765b01090 :+1:. At first was unsure about introducing a global `g_max_datacarrier_bytes`, then realized it was moved into policy in fa727f139c38e24fa23a6455b0b274177b41747c.

glozow reviewed (commented):
ACK fa727f139c38e24fa23a6455b0b274177b41747c

Lightly reviewed and tested each commit, nice move and cleanup. Would like at least another reviewer to take a look.

darosior reviewed (commented):
utACK fa727f139c38e24fa23a6455b0b274177b41747c

MarcoFalke commented on src/rpc/net.cpp:
```diff
@@ -645,8 +645,11 @@ static RPCHelpMan getnetworkinfo()
-    obj.pushKV("relayfee",      ValueFromAmount(::minRelayTxFee.GetFeePerK()));
-    obj.pushKV("incrementalfee", ValueFromAmount(::incrementalRelayFee.GetFeePerK()));
+    if (node.mempool) {
+        // Those fields can be deprecated, to be replaced by the getmempoolinfo fields
```
Personally I think it is fine to keep them forever, unless there is a strong enough reason to remove them. But this seems to be subjective.

jonatack replied:
Wondered at first look if launching bitcoind with `-blocksonly` would make these fields optional but doesn't appear to be the case (and checked that `bitcoin-cli -rpcwait getnetworkinfo` on a -blocksonly startup returns the fields).

glozow replied:
Note that -blocksonly does not mean your node doesn't create a mempool.

Commit:
sort after scripted-diff

Commit:
scripted-diff: Move mempool_args to src/node

It is part of the node library. Also, it won't be moved to the kernel
lib, as it will be pruned of ArgsManager.

-BEGIN VERIFY SCRIPT-
 # Move module
 git mv src/mempool_args.cpp src/node/
 git mv src/mempool_args.h   src/node/
 # Replacements
 sed -i 's:mempool_args\.h:node/mempool_args.h:g'     $(git grep -l mempool_args)
 sed -i 's:mempool_args\.cpp:node/mempool_args.cpp:g' $(git grep -l mempool_args)
 sed -i 's:MEMPOOL_ARGS_H:NODE_MEMPOOL_ARGS_H:g'      $(git grep -l MEMPOOL_ARGS_H)
-END VERIFY SCRIPT-

MarcoFalke commented on src/script/standard.h:
```diff
@@ -39,15 +38,6 @@ class CScriptID : public BaseHash<uint160>
- * Maximum size of TxoutType::NULL_DATA scripts that this node considers standard.
- * If nullopt, any size is nonstandard.
- */
-extern std::optional<unsigned> g_max_datacarrier_bytes;
```
Thx, fixed.

MarcoFalke commented:
Fixed doc typos and added two commits (let me know if I should drop them)

ryanofsky commented on src/mempool_args.h:
```diff
@@ -5,18 +5,23 @@
  * @param[in,out] mempool_opts The MemPoolOptions to modify according to \p argsman.
  */
-void ApplyArgsManOptions(const ArgsManager& argsman, kernel::MemPoolOptions& mempool_opts);
+[[nodiscard]] std::optional<bilingual_str> ApplyArgsManOptions(const ArgsManager& argsman, const CChainParams& chainparams, kernel::MemPoolOptions& mempool_opts);
```
In commit "Return optional error from ApplyArgsManOptions" (fa468bdfb62dec286cb977db78d3e47b64dafeba)

This seems good for now, but after #25665 this could return `util::Result<void>` instead of `std::optional<bilingual_str>`

MarcoFalke replied:
Yes, there are also a few other places. See `git grep 'std::optional<bilingual_str>'`.

ryanofsky commented on src/test/script_p2sh_tests.cpp:
```diff
@@ -18,6 +18,11 @@
 #include <boost/test/unit_test.hpp>
 
 // Helpers:
+bool IsStandardTx(const CTransaction& tx, std::string& reason)
```
In commit "Remove ::IsStandardTx(tx, reason) alias" (fa8a7f01fe1b6db98097021276ed5d929faadbec)

Would be good to declare this static to make sure nobody outside this file can call this

MarcoFalke replied:
Good point. Will do on the next push or as a follow-up.

MarcoFalke replied:
Fixed in #25772

ryanofsky reviewed (commented):
Code review ACK ddddd6913b1bdee1cad89a32d363306ea1f7b8d7

ariard commented on src/node/mempool_args.cpp:
```diff
@@ -0,0 +1,99 @@
+        mempool_opts.max_datacarrier_bytes = std::nullopt;
+    }
+
+    mempool_opts.require_standard = !argsman.GetBoolArg("-acceptnonstdtxn", !chainparams.RequireStandard());
```
could be called `require_txstandard` we might have standard package in the future, good to dissociate

MarcoFalke replied:
`require_standard` is applied to both txs and packages, so I think the current name is fine, no?

glozow replied:
I also think package standardness would be included with `-acceptnonstdtxn` anyway, can't really think of a reason not to. Feel free to reopen if something changes, we can cross this bridge if/when we come to it.

ariard commented on src/wallet/fees.cpp:
```diff
@@ -87,7 +87,7 @@ CFeeRate GetDiscardRate(const CWallet& wallet)
     // Don't let discard_rate be greater than longest possible fee estimate if we get a valid fee estimate
     discard_rate = (discard_rate == CFeeRate(0)) ? wallet.m_discard_rate : std::min(discard_rate, wallet.m_discard_rate);
-    // Discard rate must be at least dustRelayFee
+    // Discard rate must be at least dust relay feerate
```
`m_dust_relay_feerate` we do so for other variables doc refs make the code more greppable

MarcoFalke replied:
`wallet.m_dust_relay_feerate` doesn't exist, only `wallet.chain().relayDustFee()`. Happy to switch to `relayDustFee` or similar, or just remove the comment completely, as the code is self-explanatory. Just let me know.

ariard reviewed (approved):
Light Code Review ACK ddddd69

MarcoFalke commented on src/kernel/mempool_options.h:
```diff
@@ -33,6 +38,20 @@ struct MemPoolOptions {
+     */
+    std::optional<unsigned> max_datacarrier_bytes{DEFAULT_ACCEPT_DATACARRIER ? std::optional{MAX_OP_RETURN_RELAY} : std::nullopt};
+    bool permit_bare_multisig{DEFAULT_PERMIT_BAREMULTISIG};
+    bool require_standard{true};
```
All options are documented in the manpage. This one under `-acceptnonstdtxn`. Happy to review a docs pull if there is need for one and someone creates one.

glozow commented:
re ACK ddddd69

changes since last review:

- changed comment from dustRelayFee to dust relay feerate
- changed comment from nMaxDatacarrierBytes to -datacarriersize for https://github.com/bitcoin/bitcoin/pull/25648#discussion_r935467794
- moved mempool_args to src/node and sorted includes
